README.md
[![NPM](https://nodei.co/npm/super-loop.png)](https://www.npmjs.com/package/super-loop)

# Super Loop

Easy programming interface for producer-consumer problem and enhanced loop, leveraging nodejs stream. 

It provides the following features: 
* repeats cycle for `n` times;
* repeats cycle until future moment;
* repeats cycle until "ender" function returns true;
* supports tps controlling;
* supports backpressure;
* supports max concurrency configuration;

## Why

Asynchronous event-based processing is quite common in modern microservice architecture as well as data ETL. Here are some typical use cases: 

* Consume SQS messages at given max tps and max concurrency to fully utilize system capacity (e.g. CPU);
* Read upstream events and persist them to a NoSQL db at fast and controlled pace;
* Purge NoSQL db records selectively at fast and controlled pace;
* Process CSV files and call REST API at fast and controlled pace;
* Performance-test HTTP endpoints with customized data, i.e. a simplified version of K6;


The goal of this library is to provide easy programming interfaces for those common and popular use cases, so as to increase developer efficiency. 


## Installation

```shell
npm i --save super-loop
```

## Usage Examples

### Example - producer consumer problem

```js
const SuperLoop =  require('super-loop');
const stats = require('measured-core').createCollection();

async function main() {
    const intv = setInterval(function () {
        console.log(JSON.stringify(stats, null, 2));
    }, 1000);

    try {
        const producer = () => {
            stats.meter('producerTps').mark();
            return [{name: 'John'}, {name: 'Alex'}]
        }

        const consumer = (data) => {
            stats.meter('consumerTps').mark();
            // processing ...
            //console.log(data);
        }

        const loop = new SuperLoop();
        const lastFor = 120_000;

        loop.on('warn', (err) => {
            console.error(err);
        });

        await loop.producedBy(producer)
            .consumedBy(consumer)
            .concurrency(200)
            .rate(1000)
            //.repeat(100)
            .until(Date.now() + lastFor)
            .exec();

        console.log('loop ends')

    } catch (e) {
        console.error('something went wrong', e)
    }
    clearInterval(intv);
}

main().catch(console.error)
```

### More examples

* [CSV Processing](https://github.com/simple-dev-tools/super-loop/blob/main/examples/csv_processor.js)
* [Enhanced for/while Loop](https://github.com/simple-dev-tools/super-loop/blob/main/examples/simple_loop.js)
* [Performance-test your HTTP endpoints](https://github.com/simple-dev-tools/super-loop/blob/main/examples/performance_test1.js)


## API

Super Loop API design follows [Fluent Interface](https://en.wikipedia.org/wiki/Fluent_interface).  All methods return `this`, except `exec` which kicks off the execution of the loop.

### consumedBy(func)

Configure consumer function. 

Arguments: 
* `func` is a function that take nodejs stream chunk as param, and returns nothing. e.g. `(data) => { console.log(data) }` 

### producedBy(func)

Configure producer function.

Arguments: 
* `func` is a function that takes no param, but returns an Array of data chunks. e.g. `() => [1, 2, 'a', 'b']`

### invoke(func)

Alias to `consumedBy`. 

### endedBy(func)

Configure ender function. When the ender function returns `true`, the loop ends. e.g. `() => false`

Arguments: 
* `func` is a function that takes no param, but returns a boolean.  

### concurrency(maxC)

Configure max concurrency for consumer function.

Arguments: 
* `maxC` is the max concurrecy the consumer function would run at. 

### rate(tps)

Configure max tps for consumer function.

Arguments: 
* `tps` is the max tps the consumer function would run at. 

### until(endTime)

Configure timestamp (in `ms`) when the loop should stop. 

Arguments: 
* `endTime` timestamp in `ms`

### repeat(times)

Configure max repeats the producer function should be called. After max repeats are reached, loop ends. 

Arguments: 
* `times` max repeats

### pipeFrom(upstream) 

Configure custom upstream rather than using super-loop internal `Readable` as upstream. 

A good example is processing file stream.

Arguments:
* `upstream` Transform or Readable stream 

### pipeTo(downstream) 

Configure custom downstream to continue streaming.

Arguments:
* `downstream` Transform or Writable stream 

### exec() 

Start the loop for execution, until one of the ending conditions is met. The ending conditions are specified by `endedBy`, `until`, `repeat`. 

Returns:

* __Error__ if critical error happens in the internal nodejs streams

### Events

* `warn`. The loop emits non-critical errors when processing individual data chunks. Here is an example to catch them and log them properly: 

```js
loop.on('warn', (err) => {
    myLogger.error(err);
});
```


## License
super-loop is licensed under the MIT license.


Changelog.md
All notable changes to this project will be documented in this file.

#### [0.3.1] - 18-12-2022
- added examples for performance test scenario

#### [0.2.0] - 07-05-2022
- add downstream support

#### [0.2.0] - 02-05-2022
- add custom upstream support
- better API docs

#### [0.1.0] - 01-05-2022
- initial setup


__tests__/loop.test.js
const SuperLoop = require('../src/superloop');

describe('SuperLoop unit test', () => {

    it('#invoke function for x time', async () => {
        const loop = new SuperLoop();
        let counter = 0;
        const f = () => {
            counter += 1;
        }

        await loop.invoke(f).repeat(5).exec();
        expect(counter).toBe(5);
    });

    it('#invoke function until specified timestamp', async () => {
        const f = () => {
            // do nothing
        }
        const loop = new SuperLoop();
        const startTime = Date.now();
        const lastForSeconds = 2_000; // for 2 seconds
        await loop.invoke(f)
            .concurrency(100)
            .rate(100)
            .until(Date.now() + lastForSeconds)
            .exec();
        const actualTimeElpased = Date.now() - startTime;
        expect(Math.abs(actualTimeElpased - lastForSeconds)).toBeLessThan(500); // all 500ms room
    });

    it('#produce x times, consume all', async () => {
        const loop = new SuperLoop();
        let counter = 0;
        const producer = () => {
            return [{name: 'John'}, {name: 'Alex'}];
        }
        const consumer = (data) => {
            counter += 1;
        }

        await loop.producedBy(producer).consumedBy(consumer).repeat(5).exec();
        expect(counter).toBe(10);
    });

    it('#produce/consume until specified timestamp', async () => {
        const producer = () => {
            return [{name: 'John'}, {name: 'Alex'}];
        }
        const consumer = (data) => {
            // do nothing
        }
        const loop = new SuperLoop();
        const startTime = Date.now();
        const lastForSeconds = 2_000; // for 2 seconds
        await loop.
            producedBy(producer)
            .consumedBy(consumer)
            .concurrency(100)
            .rate(100)
            .until(Date.now() + lastForSeconds)
            .exec();
        const actualTimeElpased = Date.now() - startTime;
        expect(Math.abs(actualTimeElpased - lastForSeconds)).toBeLessThan(500); // all 500ms room
    });


    it('end loop if enderFunc returns true', async () => {
        const producer = () => {
            return [{name: 'John'}, {name: 'Alex'}];
        }
        let counter = 0;
        const consumer = (data) => {
            ++counter
        }

        const ender = () => {
            return counter > 8;
        }

        const loop = new SuperLoop();
        const startTime = Date.now();
        const lastForSeconds = 2_000; // for 2 seconds
        await loop.
            producedBy(producer)
            .consumedBy(consumer)
            .concurrency(100)
            .rate(100)
            .endedBy(ender)
            .exec();

        expect(counter).toBe(10);
        
    });

});


__tests__/csv_upstream.test.js
const csv = require('csv-parser');
const fs = require('fs');
const SuperLoop = require('../src/superloop');

describe('SuperLoop custom Readable as stream', () => {

    it('#process csv file', async () => {

        const upstream = fs.createReadStream(`${__dirname}/small_data.csv`).pipe(csv())

        const loop = new SuperLoop();
        let counter = 0;
        const f = () => {
            counter += 1;
        }

        await loop.pipeFrom(upstream).invoke(f).exec();
        expect(counter).toBe(3);
        
    });

})

__tests__/csv_stream_pipe.test.js
const csv = require('csv-parser');
const { Writable } = require('stream')
const fs = require('fs');
const SuperLoop = require('../src/superloop');

describe('SuperLoop custom Readable as stream', () => {

    it('#process csv file', async () => {

        const upstream = fs.createReadStream(`${__dirname}/small_data.csv`).pipe(csv())

        const loop = new SuperLoop();
        let counter = 0;
        const f = () => {
            counter += 1;
        }

        await loop.pipeFrom(upstream).invoke(f).exec();
        expect(counter).toBe(3);

    });

    it('#process csv file and write to downstream', async () => {

        const upstream = fs.createReadStream(`${__dirname}/small_data.csv`).pipe(csv())

        const loop = new SuperLoop();
        let counter = 0;

        const f = (rawEvent) => {
            counter += 1;
            return {
                data: {
                    index: rawEvent.Index,
                    height: rawEvent.Height,
                    weight: rawEvent.Weight
                }
            }
        }


        const wResults = [];
        const writer = new Writable({
            objectMode: true,
            write(data, _, done) {
                wResults.push(data)
                done()
            }
        })

        await loop.pipeFrom(upstream).invoke(f).pipeTo(writer).exec();
        expect(counter).toBe(3);
        expect(wResults).toMatchSnapshot()

    });

})

src/streams.js
const { RateLimiterMemory } = require('rate-limiter-flexible')
const { Transform, Readable } = require('stream')
const { promisify } = require('util')

const sleep = promisify(setTimeout)

class SimpleLoopReadable extends Readable {

  constructor(options) {
    super(options)
    this._suspended = false
    this.endTimestamp = options.endTimestamp || Infinity
    this.maxRepeat = options.maxRepeat || Infinity
    this.messageProducerFunc = options.messageProducerFunc
    this.streamEnderFunc = options.streamEnderFunc
    this.loggingEmitter = options.loggingEmitter
    this.start()
  }

  _read() {
    if (this._suspended) {
      this.resumeStreaming()
    }
  }

  resumeStreaming() {
    this._suspended = false
  }

  pauseStreaming() {
    this._suspended = true
  }

  start() {
    this.counter = 0
    const loop = () => {
      setImmediate(async () => {

        // precheck
        try {
          //console.log('time diff', Date.now() - this.endTimestamp, Date.now() >= this.endTimestamp );
          if (Date.now() >= this.endTimestamp || this.counter >= this.maxRepeat) {
            this.push(null)
            return
          }

          if (this.streamEnderFunc && this.streamEnderFunc()) {
            this.push(null)
            return
          }
        } catch (err) {
          err.message = 'SimpleLoopReadable Precheck Error -- ' + err.message
          this.loggingEmitter.emit('warn', err)
        }

        if (!this._suspended) {
          try {
            // messageProduceFunc should return messages array
            this.counter += 1
            const messages = this.messageProducerFunc ? await this.messageProducerFunc() : [{ index: this.counter }]

            //console.log('producer', messages[0].index);
            for (let i = 0; i < messages.length; i++) {
              const message = messages[i]
              const canContinue = this.push(message)
              // keep pushing until entire array is pushed to downstream
              if (!canContinue && i === messages.length - 1) {
                this.pauseStreaming()
                await sleep(100)
              }
            }

          } catch (err) {
            err.message = 'SimpleLoopReadable Error -- ' + err.message
            this.loggingEmitter.emit('warn', err)
          }
        }
        return loop()
      })
    }
    loop()
  }
}


function StreamThrottler(tps = 10) {
  const rateLimiter = new RateLimiterMemory({
    points: tps,
    duration: 1
  })

  function process(
    self, data, encoding, done
  ) {
    rateLimiter.consume('#super-loop#streaming_util.js#throtter', 1)
      .then(() => done(null, data))
      .catch((e) => {
        if (e.msBeforeNext) {
          setTimeout(() => {
            try {
              process(
                self, data, encoding, done
              )
            } catch (err) {
              err.message = 'StreamThrotter Error -- ' + err.message
              done(err)
            }
          }, e.msBeforeNext)
        } else {
          done(e)
        }
      })
  }

  return new Transform({
    objectMode: true,
    transform(data, encoding, done) {
      process(
        this, data, encoding, done
      )
    }
  })
}

module.exports = {
  StreamThrottler,
  SimpleLoopReadable
}


src/superloop.js
const { promisify } = require('util')
const EventEmitter = require('events')
const { pipeline } = require('stream')
const t2c = require('through2-concurrent')

const Streams = require('./streams')

class SuperLoop extends EventEmitter {

  constructor() {
    super()
    this.maxConcurrency = 100
    this.maxTps = 50
    this.maxRepeat = Infinity
    this.endTimestamp = Infinity
    this.producerFunc = null // function should tak no arg, and return Array
    this.consumerFunc = () => { }
    this.enderFunc = () => false // function return true or false
    this.upstream = null
    this.downstream = null
    this.mapperFuncs = []
  }

  consumedBy(func) {
    return this.invoke(func)
  }

  producedBy(func) {
    this.producerFunc = func
    return this
  }

  invoke(func) {
    this.consumerFunc = func
    return this
  }

  endedBy(func) {
    this.enderFunc = func
    return this
  }

  concurrency(maxC) {
    this.maxConcurrency = maxC
    return this
  }

  rate(tps) {
    this.maxTps = tps
    return this
  }

  until(endTime) {
    this.endTimestamp = endTime
    return this
  }

  repeat(times) {
    this.maxRepeat = times
    return this
  }

  pipeFrom(up) {
    this.upstream = up
    return this
  }

  pipeTo(down) {
    this.downstream = down
    return this
  }

  async exec() {
    const _func = this.consumerFunc
    const _loop = this
    let forcePipeStop = false
    
    try {

      if (!this.upstream) {
        this.upstream = new Streams.SimpleLoopReadable({
          objectMode: true,
          maxRepeat: this.maxRepeat,
          endTimestamp: this.endTimestamp,
          messageProducerFunc: this.producerFunc,
          streamEnderFunc: () => this.enderFunc() || forcePipeStop,
          loggingEmitter: _loop
        })
      }

      const streams = [
        this.upstream,
        Streams.StreamThrottler(this.maxTps),
        t2c.obj({
          objectMode: true,
          maxConcurrency: this.maxConcurrency
        }, async function (data, _, cb) {
          try {
            const result = await _func(data)
            if (_loop.downstream) { // only if downstream exists
              this.push(result)
            }
          } catch (err) {
            err.message = 'SuperLoop#exec Error -- ' + err.message
            _loop.emit('warn', err)
          } finally {
            cb()
          }
        })
      ]
      
      if (this.downstream) {
        streams.push(this.downstream)
      }

      await promisify(pipeline)(
        ...streams
      )

    } catch (pipeError) {
      forcePipeStop = true
      throw pipeError
    }

  }
}

module.exports = SuperLoop


examples/csv_processor.js
const csv = require('csv-parser');
const fs = require('fs');

const SuperLoop = require('../index');
const stats = require('measured-core').createCollection();

async function main() {
    const intv = setInterval(function () {
        console.log(JSON.stringify(stats, null, 2));
    }, 1000);

    try {

        const upstream = fs.createReadStream('./examples/csv_sample.csv').pipe(csv())

        const consumer = (data) => {
            stats.meter('consumerTps').mark();
            // processing ...
            // console.log(data)
            // calling another API to process csv records
            //throw new Error('processing error');
        }

        const loop = new SuperLoop();
        const lastFor = 120_000;

        loop.on('warn', (err) => {
            console.error(err);
        });

        await loop.pipeFrom(upstream)
            .consumedBy(consumer)
            .concurrency(200)
            .rate(1000)
            //.repeat(100)
            .until(Date.now() + lastFor)
            .exec();

        console.log('loop ends')

    } catch (e) {
        console.error('something went wrong', e)
    }
    clearInterval(intv);
}

main().catch(console.error)



examples/loop_example1.js
const SuperLoop = require('../index');
const stats = require('measured-core').createCollection();

async function main() {
    const intv = setInterval(function () {
        console.log(JSON.stringify(stats));
    }, 1000);

    try {
        const f = (data) => {
            stats.meter('requestsPerSecond').mark();
            // processing ...
            //console.log(data);
            //throw new Error('processing error');
        }
        const loop = new SuperLoop();
        const lastFor = 120_000;

        loop.on('warn', (err) => {
            console.error(err);
        });

        await loop.invoke(f)
            .concurrency(1000)
            .rate(10000)
            //.repeat(100)
            .until(Date.now() + lastFor)
            .exec();

        console.log('loop ends')

    } catch (e) {
        console.error('something went wrong', e)
    }
    clearInterval(intv);
}

main().catch(console.error)



examples/performance_test1.js
// TEST REST ENDPOINT, where it's service is hosted in GCP Cloud Run

const { GoogleAuth } = require('google-auth-library');
const SuperLoop = require('super-loop');
const { Readable } = require('node:stream');
const stats = require('measured-core').createCollection();
const histogram = stats.histogram('latency');

const BASE_URL = process.env.API_BASE_URL;

// decode private key if stored in base64 form
const SA_KEY = JSON.parse(
    Buffer.from(process.env.GCP_IAM_KEY, 'base64').toString(
        'ascii',
    ),
);

const auth = new GoogleAuth({
    credentials: SA_KEY
});

let client = null;

async function invoke(url, body) {
    // client can be reused
    if (client == null) {
        client = await auth.getIdTokenClient(process.env.TARGET_AUDIENCE); // set baseUrl as targetAudience
    }
    return client.request({
        method: 'POST',
        url,
        data: {
            ...body,
        },
    });
}

let startIndex = 99900000000;

async function send_message(offset) {
    const testTarget = `${BASE_URL}/send_message`;

    const index = startIndex + offset;

    const txId = `pt-tx-${index}`;
    let userId = `${index}`;

    if (index % 2000 == 0) { // interleaving some special data in the feed
        userId = '8888888'
    }

    try {
        const result = await invoke(testTarget, {
            "transactionId": txId,
            "userId": userId,
            "topic": "MESSAGE_TYPE_1",
            "title": "G'day mate!",
            "body": "This is a good day!",
            "eventTimestamp": "2022-02-10T05:35:31.165Z",
            "correlationId": "12333abc",
            "pt": true
        });
        console.log(JSON.stringify(result.data));
    } catch (err) {
        if (err?.response?.data?.code >= 400 ||  err?.response?.data?.code < 500) {
            console.log(JSON.stringify(err?.response?.data));
        } else {
            console.error('failed to send message', err);
        }
    }
}


// pumping requests using super-loop
// measurement is done using measured-core
(async function () {
    const intv = setInterval(function () {
        console.log(JSON.stringify(stats, null, 2));
        //console.log(histogram.toJSON());
    }, 1000);

    const sloop = new SuperLoop();
    sloop.on('warn', (e) => {
        console.warn(e);
    });
    
    await sloop.pipeFrom(Readable.from(Array.from(Array(1_000_000).keys())))
        .rate(75)
        .concurrency(200)
        .invoke(async function (data) {
            stats.meter('throughput').mark();
            const ts_from = Date.now();
            await send_message(data);
            const latency = Date.now() - ts_from;
            histogram.update(latency); 
        })
        .exec();

    clearInterval(intv);

}()).catch(console.error);


examples/producer_consumer.js
const SuperLoop = require('../index');
const stats = require('measured-core').createCollection();

async function main() {
    const intv = setInterval(function () {
        console.log(JSON.stringify(stats, null, 2));
    }, 1000);

    try {
        const producer = () => {
            stats.meter('producerTps').mark();
            // processing ...
            //console.log(data);
            //throw new Error('processing error');
            return [{name: 'John'}, {name: 'Alex'}]
        }

        const consumer = (data) => {
            stats.meter('consumerTps').mark();
            // processing ...
            //console.log(data);
            //throw new Error('processing error');
        }

        const loop = new SuperLoop();
        const lastFor = 120_000;

        loop.on('warn', (err) => {
            console.error(err);
        });

        await loop.producedBy(producer)
            .consumedBy(consumer)
            .concurrency(200)
            .rate(1000)
            //.repeat(100)
            .until(Date.now() + lastFor)
            .exec();

        console.log('loop ends')

    } catch (e) {
        console.error('something went wrong', e)
    }
    clearInterval(intv);
}

main().catch(console.error)

